addons/mail/static/src/js/document_viewer.js

```diff
@@ -0,0 +1,289 @@
+odoo.define('mail.DocumentViewer', function (require) {
+"use strict";
+
+var core = require('web.core');
+var Widget = require('web.Widget');
+
+var QWeb = core.qweb;
+
+var SCROLL_ZOOM_STEP = 0.1;
+var ZOOM_STEP = 0.5;
+
+var DocumentViewer = Widget.extend({
+    template: "DocumentViewer",
+    events: {
+        'click .o_download_btn': '_onDownload',
+        'click .o_viewer_img': '_onImageClick',
+        'click .o_viewer_video': '_onVideoClick',
+        'click .move_next': '_onNext',
+        'click .move_previous': '_onPrevious',
+        'click .o_zoom_in': '_onZoomIn',
+        'click .o_zoom_out': '_onZoomOut',
+        'click .o_close_btn, .o_viewer_img_wrapper': '_onClose',
+        'click .o_print_btn': '_onPrint',
+        'DOMMouseScroll .o_viewer_content': '_onScroll',    // Firefox
+        'mousewheel .o_viewer_content': '_onScroll',        // Chrome, Safari, IE
+        'keydown': '_onKeydown',
+        'mousedown .o_viewer_img': '_onStartDrag',
+        'mousemove .o_viewer_content': '_onDrag',
+        'mouseup .o_viewer_content': '_onEndDrag'
+    },
+    /**
+     * When initialize document viewer get image and video type attachment from thread
+     * and store it in instance so we can traversal attachment on next previous
+     *
+     * keep track of currently loaded attachment by 'active_attachment'
```

Comment:
I'd say:
The documentViewer takes an array of attachment IDs in argument. Documents that are not of type
image or video are filtered out.

Reply:
Just realized it's actually an array of objects describing the attachments, so the variable name is bad